### PR TITLE
Add support for streaming stdout/stderr from Child invocations

### PR DESCRIPTION
**Pull request by brianmario**
As the title says, this adds support for passing a block for receiving chunks of output stdout/stderr as they're being read.

I'm not _super_ happy with how the API turned out, and I could probably add a few more tests - so suggestions are definitely welcome.

@tmm1 @piki @carlosmn @simonsj @vmg @scottjg in case either of you have time to review this 🙏

### CODE REVIEW

File: lib/posix/spawn.rb [hunk 1]
```diff
@@ -337,6 +337,10 @@ class MaximumOutputExceeded < StandardError
     class TimeoutExceeded < StandardError
     end
 
+    # Exception raised when output streaming is aborted early.
+    class Aborted < StandardError
```
Comment (brianmario) [inline lib/posix/spawn.rb]: This could maybe be `CallerAborted` or something more specific?

File: lib/posix/spawn/child.rb [hunk 2]
```diff
@@ -244,14 +251,35 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
...
+                  @out << chunk
+                end
+              else
+                if @streaming && @stderr_block
+                  abort = !!@stderr_block.call(chunk)
+                else
+                  @err << chunk
+                end
+              end
+            end
```
Comment (brianmario) [inline lib/posix/spawn/child.rb]: This whole block is pretty gross, but the alternative may involve being tricky (and less readable?) with Ruby.

File: lib/posix/spawn/child.rb [hunk 3]
```diff
@@ -244,14 +251,35 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
...
+
+            if @streaming && abort
+              raise Aborted
```
Comment (brianmario) [inline lib/posix/spawn/child.rb]: I don't love raising here, but it enforces proper cleanup (and killing the subprocess) up on https://github.com/rtomayko/posix-spawn/blob/0c02f3337575a9fcdd1ba6135882afcbb0aaf290/lib/posix/spawn/child.rb#L168.

File: lib/posix/spawn/child.rb (hunk 2 above)
Comment (mhagger) [inline lib/posix/spawn/child.rb]: This might be the "tricky" Ruby you're talking about, but it seems to me that when streaming is not requested, you could set `@stdout_block` anyway, to

    Proc.new do |chunk|
      @out << chunk
      false
    end

(like you do for the tests) and do the equivalent for `@stderr_block`. Then you could avoid the inner conditionals here.

To shrink the code even further, you could set

    @blocks = { stdout => @stdout_block, stderr => @stderr_block }

(in which case you wouldn't even need `@stdout_block` and `@stderr_block` anymore, but you get the idea) then this whole processing code could become

    if @blocks[fd].call(chunk)
      raise Aborted
    end

File: lib/posix/spawn/child.rb [hunk 4]
```diff
@@ -244,14 +251,35 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
...
+              else
+                if @streaming && @stderr_block
+                  abort = !!@stderr_block.call(chunk)
+                else
+                  @err << chunk
+                end
+              end
+            end
+
+            if @streaming && abort
```
Comment (mhagger) [inline lib/posix/spawn/child.rb]: I think `@streaming &&` is redundant here, since when `@streaming` is not set, `abort` retains its initial value, `false`.

File: test/test_child.rb [hunk 5]
```diff
@@ -224,6 +224,65 @@ def test_utf8_input_long
...
+    end
+
+    input = "hello!"
+    assert_raises POSIX::Spawn::Aborted do
+      p = Child.new('ls', '-?', :streams => {
+        :stderr => stderr_stream
+      })
+    end
+  end
+
```
Comment (mhagger) [inline test/test_child.rb]: There are no tests that involve reading more than one `BUFSIZE` worth of output, or reading from both stdout and stderr. Those might be worthwhile additions.

File: lib/posix/spawn/child.rb [hunk 6]
```diff
@@ -196,16 +209,16 @@ def exec!
...
         max = nil if max && max <= 0
-        @out, @err = '', ''
 
         # force all string and IO encodings to BINARY under 1.9 for now
-        if @out.respond_to?(:force_encoding) and stdin.respond_to?(:set_encoding)
+        if @stdout_buffer.respond_to?(:set_encoding)
+          bin_encoding = Encoding::BINARY
           [stdin, stdout, stderr].each do |fd|
-            fd.set_encoding('BINARY', 'BINARY')
+            fd.set_encoding(bin_encoding, bin_encoding)
```
Comment (mclark) [inline lib/posix/spawn/child.rb]: Can we do this on a per `fd` basis?
```ruby
bin_encoding = Encoding::BINARY
[stdin, stdout, stderr].each do |fd|
  fd.set_encoding(bin_encoding, bin_encoding) if fd.respond_to?(:set_encoding)
end
```

Also, are we intentionally dropping the `force_encoding` calls on `stdout` and `stderr` here?

File: lib/posix/spawn/child.rb [hunk 7]
```diff
@@ -196,16 +209,16 @@ def exec!
...
           end
-          @out.force_encoding('BINARY')
-          @err.force_encoding('BINARY')
-          input = input.dup.force_encoding('BINARY') if input
+          @stdout_buffer.set_encoding(bin_encoding)
```
Comment (mclark) [inline lib/posix/spawn/child.rb]: Are these duplicate calls to the above intentional? We really can't assume these objects respond to these methods any more.

File: lib/posix/spawn/child.rb [hunk 8]
```diff
@@ -244,14 +257,27 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
...
             rescue Errno::EAGAIN, Errno::EINTR
             rescue EOFError
               readers.delete(fd)
               fd.close
             end
+
+            abort = false
+            if chunk
+              if fd == stdout
+                abort = (@stdout_buffer.write(chunk) == 0)
```
Comment (mclark) [inline lib/posix/spawn/child.rb]: I don't feel like this is a safe way to test for aborting the operation. The output object could be simply refusing to write the current chunk but is not done consuming the stream.
Why not use an exception for this test instead? If the consumer raises `Posix::Spawn::Aborted` then we clearly know to abort.

File: lib/posix/spawn/child.rb [hunk 9]
```diff
@@ -262,12 +288,10 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
           end
 
           # maybe we've hit our max output
-          if max && ready[0].any? && (@out.size + @err.size) > max
+          if max && ready[0].any? && (@stdout_buffer.size + @stderr_buffer.size) > max
```
Comment (mclark) [inline lib/posix/spawn/child.rb]: again, we can't assume there is a `#size` method on these objects...

I think we should probably keep a local count of the bytes we have written instead of calling `#size` anyway. We can't trust these objects any more as they could be anything.

File: test/test_child.rb (hunk 5 above)
Comment (mclark) [inline test/test_child.rb]: We should also add a test for passing in a minimal custom object, just to ensure the interface contract is maintained. May be a good time to use a spy.

File: lib/posix/spawn.rb (hunk 1 above)
Comment (mclark) [inline lib/posix/spawn.rb]: I'm fine with either.

**Comment by mclark** [comment]
Oh right, we also need some thorough docs on this once we've nailed down the exact :duck: interface we are using

File: lib/posix/spawn/child.rb [hunk 10]
```diff
@@ -244,14 +257,27 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
 
           # read from stdout and stderr streams
           ready[0].each do |fd|
-            buf = (fd == stdout) ? @out : @err
+            chunk = nil
             begin
-              buf << fd.readpartial(BUFSIZE)
+              chunk = fd.readpartial(BUFSIZE)
```
Comment (peff) [inline lib/posix/spawn/child.rb]: If I'm understanding this right, the old code would always append directly into the final buffer, whereas this one reads a chunk and then appends that chunk to the buffer. Not knowing anything about how Ruby operates under the hood, is this a potential performance problem? It _should_ just be an extra `memcpy` in the worst case, but I recall that we've hit bottlenecks on reading into Ruby before. I suspect those were mostly about arrays and not buffers, though (e.g., reading each line into its own buffer can be slow).

Comment (brianmario) [inline lib/posix/spawn/child.rb]: I could be wrong here (cc @tenderlove) but I'm pretty sure the previous code would actually create a new string (down inside `readpartial`), then _that_ string would have been appended to `buf`. Which would require potentially resizing that buffer first, then the `memcpy`.

This new code just keeps that first string as a local var first, so we can later determine where to write it. In the default case we're using a `StringIO` so the result is essentially the same as before (potential buffer resize then a copy). Though iirc we saw some pretty significant speedups by using an array to keep track of chunks, then calling `join` at the end when it was all needed. The reason for that is because it avoids the reallocation of the resulting buffer as we're reading, and instead allows `join` to allocate a buffer exactly the size that's needed then copying all the chunks in to it.

Basically this (the old way):

``` ruby
buffer = ""
buffer << "one,"
buffer << "two,"
buffer << "three"

return buffer
```

vs this (the array optimized version I just mentioned):

``` ruby
buffer = []
buffer << "one,"
buffer << "two,"
buffer << "three"

# buffer is an array with 3 elements at this point
# and this join call figures out how big all of the strings inside are, then creates a single buffer to append them to.
return buffer.join
```

Using that approach efficiently may change the API contract here slightly though...

Comment (peff) [inline lib/posix/spawn/child.rb]: @brianmario Ah, right, that sort of return value optimization would be pretty easy to implement, and would mean we end up with the same number of copies. Though if we're just counting `memcpy`s anyway, I suspect it doesn't matter much either way.

> The reason for that is because it avoids the reallocation of the resulting buffer as we're reading, and instead allows join to allocate a buffer exactly the size that's needed then copying all the chunks in to it.

Interesting. It sounds like appending doesn't grow the buffer aggressively in that case, because you should be able to get amortized constant-time.

Anyway. We're well out of my level for intelligent discussion of Ruby internals. The interesting result is whether reading the output of a spawned `cat some-gigabyte-file` is measurably any different. Probably not, but it's presumably easy to test.

File: lib/posix/spawn/child.rb [hunk 11]
```diff
@@ -223,6 +248,10 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
         slice_method = input.respond_to?(:byteslice) ? :byteslice : :slice
         t = timeout
 
+        streams = {stdout => @stdout_stream, stderr => @stderr_stream}
+
+        bytes_seen = 0
+        chunk_buffer = ""
```
Comment (brianmario) [inline lib/posix/spawn/child.rb]: This buffer is reused by `readpartial` below. Internally, so far as I can tell, it will be resized to `BUFSIZE` on the first call and the underlying buffer reused from then on out.

File: lib/posix/spawn/child.rb [hunk 12]
```diff
@@ -262,12 +294,10 @@ def read_and_write(input, stdin, stdout, stderr, timeout=nil, max=nil)
           end
 
           # maybe we've hit our max output
-          if max && ready[0].any? && (@out.size + @err.size) > max
+          if max && ready[0].any? && bytes_seen > max
             raise MaximumOutputExceeded
           end
         end
-
-        [@out, @err]
```
Comment (brianmario) [inline lib/posix/spawn/child.rb]: I decided to just drop returning these in an attempt at consistency since these one or both of these ivars are useless if we're streaming.

File: test/test_child.rb [hunk 13]
```diff
@@ -224,6 +224,79 @@ def test_utf8_input_long
...
+      chunk_count += 1
+      stdout_buff << chunk
+    end
+
+    limit = Child::BUFSIZE * 2
+
+    begin
+      Child.new('yes', :streams => {:stdout => stdout_stream}, :max => limit)
+    rescue POSIX::Spawn::MaximumOutputExceeded
+    end
```
Comment (brianmario) [inline test/test_child.rb]: Good call 👍

**Comment by carlosmn** [comment]
This doesn't seem to be new, but the semantics of `:max` seem rather confusing. It doesn't mean "send me at most X bytes" but rather "if the process sends over X amount, give it to me and then raise an error", which means that for a `:max` which is a multiple of the effective buffer size, you get a whole extra chunk.

While I was playing around with the added test here, I noticed that all chunks are 16kB (which seems to be the default pipe buffer size here) so the `BUFSIZE * 2` size four times the chunk size (since `yes` outputs so quickly, we always find it full) which means the output stream gets a whole extra 16kB beyond what is specified as `:max`. This seems to be semantics that are already there, but these seems really hard to plan for, if I can get an arbitrary amount of bytes beyond what I asked for.

File: lib/posix/spawn/child.rb [hunk 14]
```diff
@@ -95,6 +95,31 @@ def initialize(*args)
           @options[:pgroup] = true
         end
         @options.delete(:chdir) if @options[:chdir].nil?
+
+        @out, @err = "", ""
```
Comment (jbarnette) [inline lib/posix/spawn/child.rb]: Might want to use `String.new` to avoid breaking when someone passes `--enable-frozen-string-literal`.

File: lib/posix/spawn/child.rb [hunk 15]
```diff
@@ -95,6 +95,31 @@ def initialize(*args)
...
+
+        @stdout_stream = Proc.new do |chunk|
+          @out << chunk
+
+          true
+        end
```
Comment (jbarnette) [inline lib/posix/spawn/child.rb]:
```ruby
@stdout_stream = @out.method(:<<)
```

:trollface: 🚲 🏠

File: lib/posix/spawn/child.rb (hunk 11 above)
Comment (jbarnette) [inline lib/posix/spawn/child.rb]: Another `""` that might be better as `String.new` here.

File: lib/posix/spawn/child.rb (hunk 15 above)
Comment (brianmario) [inline lib/posix/spawn/child.rb]: No I love it! Being able to use `method` was one of the main reasons for going back to the proc-based API ;)

**Comment by arrbee** [comment]
Any plans to merge this? I'd like to use it :-D
